File: examples/procgen/c/src/procgen.c
#include <math.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <string.h>
#include <emscripten/console.h>
#define FNL_IMPL
#include "./FastNoiseLite.h"
#include "../../../../src/engine/scripting/emscripten/src/websg.h"
#include "../../../../src/engine/scripting/emscripten/src/thirdroom.h"

#define MIN(a,b) (((a)<(b))?(a):(b))
#define MAX(a,b) (((a)>(b))?(a):(b))
#define F_EPSILON 0.00000005f
#define F_EQ(x,y) (((x)*(x) - (y)*(y) <= F_EPSILON) ? 1 : 0)

struct SphereData {
  int vertex_count;
  accessor_id_t positions_accessor;
  float_t *positions;
  size_t positions_byte_length;
  mesh_id_t mesh_id;
} typedef SphereData;

// Adapted from https://github.com/dmnsgn/primitive-geometry/blob/main/src/ellipsoid.js
SphereData *generate_sphere_mesh(
  float_t radius,
  int widthSegments,
  int heightSegments,
  float_t theta,
  float_t thetaOffset,
  float_t phi,
  float_t phiOffset
) {
  size_t byte_length = 0;

  int size = (widthSegments + 1) * (heightSegments + 1);

  int indices_count = widthSegments * heightSegments * 6;
  size_t indices_byte_length = sizeof(uint16_t) * indices_count;
  uint16_t *indices = malloc(indices_byte_length);

  int positions_count = size;
  size_t positions_byte_length = sizeof(float_t) * positions_count * 3;
  float_t *positions = malloc(positions_byte_length);

  int normals_count = size;
  size_t normals_byte_length = sizeof(float_t) * normals_count * 3;
  float_t *normals = malloc(normals_byte_length);

  int uvs_count = size;
  size_t uvs_byte_length = sizeof(float_t) * uvs_count * 2;
  float_t *uvs = malloc(uvs_byte_length);

  float_t temp[3];
  int vertexIndex = 0;
  int curIndex = 0;

  for ( int iy = 0; iy <= heightSegments; iy ++ ) {

    float_t v = (float)iy / heightSegments;
    float_t t = v * theta + thetaOffset;
    float_t cosTheta = cosf(t);
    float_t sinTheta = sinf(t);

    for ( int ix = 0; ix <= widthSegments; ix ++ ) {

      float_t u = (float)ix / widthSegments;
      float_t p = u * phi + phiOffset;
      float_t cosPhi = cosf(p);
      float_t sinPhi = sinf(p);

      temp[0] = -cosPhi * sinTheta;
      temp[1] = -cosTheta;
      temp[2] = sinPhi * sinTheta;

      positions[vertexIndex * 3] = radius * temp[0];
      positions[vertexIndex * 3 + 1] = radius * temp[1];
      positions[vertexIndex * 3 + 2] = radius * temp[2];

      float_t length = sqrtf(temp[0] * temp[0] + temp[1] * temp[1] + temp[2] * temp[2]);
      float_t n = 1 / length;
      normals[vertexIndex * 3] = temp[0] * n;
      normals[vertexIndex * 3 + 1] = temp[1] * n;
      normals[vertexIndex * 3 + 2] = temp[2] * n;

      uvs[vertexIndex * 2] = u;
      uvs[vertexIndex * 2 + 1] = v;

      vertexIndex++;
    }

    if (iy > 0) {
      for (int i = vertexIndex - 2 * (widthSegments + 1); i + widthSegments + 2 < vertexIndex; i++) {
        int16_t a = i;
        int16_t b = i + 1;
        int16_t c = i + widthSegments + 1;
        int16_t d = i + widthSegments + 2;
        indices[curIndex] = a;
        indices[curIndex + 1] = b;
        indices[curIndex + 2] = c;

        indices[curIndex + 3] = c;
        indices[curIndex + 4] = b;
        indices[curIndex + 5] = d;

        curIndex += 6;
      }
    }
  }

  MeshPrimitiveProps *primitive_props = malloc(sizeof(MeshPrimitiveProps));
  primitive_props->mode = MeshPrimitiveMode_TRIANGLES;
  
  AccessorFromProps *indices_props = malloc(sizeof(AccessorFromProps));
  indices_props->component_type = AccessorComponentType_Uint16;
  indices_props->count = indices_count;
  indices_props->type = AccessorType_SCALAR;
  accessor_id_t indices_accessor = websg_world_create_accessor_from(indices, indices_byte_length, indices_props);
  primitive_props->indices = indices_accessor;

  primitive_props->attributes.items = malloc(sizeof(MeshPrimitiveAttributeItem) * 3);
  primitive_props->attributes.count = 3;

  AccessorFromProps *positions_props = malloc(sizeof(AccessorFromProps));
  positions_props->component_type = AccessorComponentType_Float32;
  positions_props->count = positions_count;
  positions_props->type = AccessorType_VEC3;
  positions_props->dynamic = true;
  accessor_id_t positions_accessor = websg_world_create_accessor_from(positions, positions_byte_length, positions_props);
  primitive_props->attributes.items[0].key = MeshPrimitiveAttribute_POSITION;
  primitive_props->attributes.items[0].accessor_id = positions_accessor;

  AccessorFromProps *normals_props = malloc(sizeof(AccessorFromProps));
  normals_props->component_type = AccessorComponentType_Float32;
  normals_props->count = normals_count;
  normals_props->type = AccessorType_VEC3;
  normals_props->normalized = true;
  normals_props->dynamic = true;
  accessor_id_t normals_accessor = websg_world_create_accessor_from(normals, normals_byte_length, normals_props);
  primitive_props->attributes.items[1].key = MeshPrimitiveAttribute_NORMAL;
  primitive_props->attributes.items[1].accessor_id = normals_accessor;

  AccessorFromProps *uvs_props = malloc(sizeof(AccessorFromProps));
  uvs_props->component_type = AccessorComponentType_Float32;
  uvs_props->count = uvs_count;
  uvs_props->type = AccessorType_VEC2;
  accessor_id_t uvs_accessor = websg_world_create_accessor_from(uvs, uvs_byte_length, uvs_props);
  primitive_props->attributes.items[2].key = MeshPrimitiveAttribute_TEXCOORD_0;
  primitive_props->attributes.items[2].accessor_id = uvs_accessor;

  MeshProps *mesh_props = malloc(sizeof(MeshProps));
  mesh_props->primitives.items = primitive_props;
  mesh_props->primitives.count = 1;

  mesh_id_t mesh_id = websg_world_create_mesh(mesh_props);

  SphereData *sphere_data = malloc(sizeof(SphereData));
  sphere_data->vertex_count = vertexIndex;
  sphere_data->positions = positions;
  sphere_data->positions_accessor = positions_accessor;
  sphere_data->positions_byte_length = positions_byte_length;
  sphere_data->mesh_id = mesh_id;

  return sphere_data;
}

fnl_state noise_state;
SphereData *sphere_data;
node_id_t sphere_node;
float_t *sphere_scale;
material_id_t beam_material;
float_t *beam_emissive_factor;
unsigned char *audio_data;

export int32_t websg_load() {
  MaterialProps *material_props = malloc(sizeof(MaterialProps));
  MaterialPbrMetallicRoughnessProps *pbr_metallic_roughness_props = malloc(sizeof(MaterialPbrMetallicRoughnessProps));
  pbr_metallic_roughness_props->base_color_factor[0] = 0;
  pbr_metallic_roughness_props->base_color_factor[1] = 0;
  pbr_metallic_roughness_props->base_color_factor[2] = 0;
  pbr_metallic_roughness_props->base_color_factor[3] = 1;
  pbr_metallic_roughness_props->metallic_factor = 0.5f;
  pbr_metallic_roughness_props->roughness_factor = 0.7f;
  material_props->pbr_metallic_roughness = pbr_metallic_roughness_props;

  material_id_t mesh_material = websg_world_create_material(material_props);
  sphere_data = generate_sphere_mesh(1, 32, 16, M_PI, 0, M_PI * 2, 0);
  websg_mesh_set_primitive_material(sphere_data->mesh_id, 0, mesh_material);

  sphere_scale = malloc(sizeof(float_t) * 3);
  sphere_scale[0] = 3;
  sphere_scale[1] = 3;
  sphere_scale[2] = 3;


  NodeProps *node_props = malloc(sizeof(NodeProps));
  node_props->translation[0] = 7;
  node_props->translation[1] = 8;
  node_props->translation[2] = -17;
  node_props->rotation[0] = 0;
  node_props->rotation[1] = 0;
  node_props->rotation[2] = 0;
  node_props->rotation[3] = 1;
  node_props->scale[0] = sphere_scale[0];
  node_props->scale[1] = sphere_scale[1];
  node_props->scale[2] = sphere_scale[2];
  node_props->mesh = sphere_data->mesh_id;
  sphere_node = websg_world_create_node(node_props);

  scene_id_t scene = websg_world_get_environment();
  websg_scene_add_node(scene, sphere_node);

  const char *beam_name = "Tube_Light_01";
  mesh_id_t beam_mesh = websg_world_find_mesh_by_name(beam_name, strlen(beam_name));
  beam_material = websg_mesh_get_primitive_material(beam_mesh, 0);
  beam_emissive_factor = malloc(sizeof(float_t) * 3);
  beam_emissive_factor[0] = 1;
  beam_emissive_factor[1] = 1;
  beam_emissive_factor[2] = 1;

  uint32_t audio_data_size = thirdroom_get_audio_data_size();
  audio_data = malloc(audio_data_size);
  noise_state = fnlCreateState();
  noise_state.noise_type = FNL_NOISE_OPENSIMPLEX2;
  noise_state.frequency = 0.01;
  noise_state.fractal_type = FNL_FRACTAL_PINGPONG;

  return 0;
}

float_t elapsed = 0;
float_t acc_noise = 0;

export int32_t websg_update(float_t dt) {
  elapsed += dt;

  thirdroom_get_audio_frequency_data(audio_data);

  float_t low_freq_avg;

  for (int i = 0; i < 128; i++) {
    low_freq_avg += audio_data[i];
  }

  low_freq_avg =  ((float)low_freq_avg / 128) / 255;

  for (int i = 0; i < sphere_data->vertex_count; i++) {
    float_t x = sphere_data->positions[i * 3];
    float_t y = sphere_data->positions[i * 3 + 1];
    float_t z = sphere_data->positions[i * 3 + 2];

    float_t length = sqrtf(x * x + y * y + z * z);
    float_t n = 1 / length;
    x = x * n;
    y = y * n;
    z = z * n;

    float_t noise = 1 + 0.5 * fnlGetNoise3D(&noise_state, x * 5 + elapsed, y * 5 + elapsed, z * 5 + elapsed);
    sphere_data->positions[i * 3] = x * noise;
    sphere_data->positions[i * 3 + 1] = y * noise;
    sphere_data->positions[i * 3 + 2] = z * noise;

    acc_noise += noise;
  }

  float_t scale = 3 * (1 + low_freq_avg);
  sphere_scale[0] = scale;
  sphere_scale[1] = scale;
  sphere_scale[2] = scale;
  websg_node_set_scale(sphere_node, sphere_scale);

  beam_emissive_factor[0] = 1;
  beam_emissive_factor[1] = 1 - low_freq_avg;
  beam_emissive_factor[2] = 1 - low_freq_avg;
  websg_material_set_emissive_factor(beam_material, beam_emissive_factor);

  websg_accessor_update_with(
    sphere_data->positions_accessor,
    sphere_data->positions,
    sphere_data->positions_byte_length
  );

  return 0;
}

File: examples/procgen/js/procgen.js
// https://github.com/mrdoob/three.js/blob/master/examples/jsm/math/ImprovedNoise.js
const _p = [
  151, 160, 137, 91, 90, 15, 131, 13, 201, 95, 96, 53, 194, 233, 7, 225, 140, 36, 103, 30, 69, 142, 8, 99, 37, 240, 21,
  10, 23, 190, 6, 148, 247, 120, 234, 75, 0, 26, 197, 62, 94, 252, 219, 203, 117, 35, 11, 32, 57, 177, 33, 88, 237, 149,
  56, 87, 174, 20, 125, 136, 171, 168, 68, 175, 74, 165, 71, 134, 139, 48, 27, 166, 77, 146, 158, 231, 83, 111, 229,
  122, 60, 211, 133, 230, 220, 105, 92, 41, 55, 46, 245, 40, 244, 102, 143, 54, 65, 25, 63, 161, 1, 216, 80, 73, 209,
  76, 132, 187, 208, 89, 18, 169, 200, 196, 135, 130, 116, 188, 159, 86, 164, 100, 109, 198, 173, 186, 3, 64, 52, 217,
  226, 250, 124, 123, 5, 202, 38, 147, 118, 126, 255, 82, 85, 212, 207, 206, 59, 227, 47, 16, 58, 17, 182, 189, 28, 42,
  223, 183, 170, 213, 119, 248, 152, 2, 44, 154, 163, 70, 221, 153, 101, 155, 167, 43, 172, 9, 129, 22, 39, 253, 19, 98,
  108, 110, 79, 113, 224, 232, 178, 185, 112, 104, 218, 246, 97, 228, 251, 34, 242, 193, 238, 210, 144, 12, 191, 179,
  162, 241, 81, 51, 145, 235, 249, 14, 239, 107, 49, 192, 214, 31, 181, 199, 106, 157, 184, 84, 204, 176, 115, 121, 50,
  45, 127, 4, 150, 254, 138, 236, 205, 93, 222, 114, 67, 29, 24, 72, 243, 141, 128, 195, 78, 66, 215, 61, 156, 180,
];

for (let i = 0; i < 256; i++) {
  _p[256 + i] = _p[i];
}

function fade(t) {
  return t * t * t * (t * (t * 6 - 15) + 10);
}

function lerp(t, a, b) {
  return a + t * (b - a);
}

function grad(hash, x, y, z) {
  const h = hash & 15;
  const u = h < 8 ? x : y,
    v = h < 4 ? y : h == 12 || h == 14 ? x : z;
  return ((h & 1) == 0 ? u : -u) + ((h & 2) == 0 ? v : -v);
}

function improvedNoise(x, y, z) {
  const floorX = Math.floor(x),
    floorY = Math.floor(y),
    floorZ = Math.floor(z);

  const X = floorX & 255,
    Y = floorY & 255,
    Z = floorZ & 255;

  x -= floorX;
  y -= floorY;
  z -= floorZ;

  const xMinus1 = x - 1,
    yMinus1 = y - 1,
    zMinus1 = z - 1;

  const u = fade(x),
    v = fade(y),
    w = fade(z);

  const A = _p[X] + Y,
    AA = _p[A] + Z,
    AB = _p[A + 1] + Z,
    B = _p[X + 1] + Y,
    BA = _p[B] + Z,
    BB = _p[B + 1] + Z;

  return lerp(
    w,
    lerp(
      v,
      lerp(u, grad(_p[AA], x, y, z), grad(_p[BA], xMinus1, y, z)),
      lerp(u, grad(_p[AB], x, yMinus1, z), grad(_p[BB], xMinus1, yMinus1, z))
    ),
    lerp(
      v,
      lerp(u, grad(_p[AA + 1], x, y, zMinus1), grad(_p[BA + 1], xMinus1, y, zMinus1)),
      lerp(u, grad(_p[AB + 1], x, yMinus1, zMinus1), grad(_p[BB + 1], xMinus1, yMinus1, zMinus1))
    )
  );
}

// Adapted from https://github.com/dmnsgn/primitive-geometry/blob/main/src/ellipsoid.js

function generateSphereMesh(radius, widthSegments, heightSegments, theta, thetaOffset, phi, phiOffset) {
  const size = (widthSegments + 1) * (heightSegments + 1);

  const indicesCount = widthSegments * heightSegments * 6;
  const indices = new Uint16Array(indicesCount);

  const positionsCount = size;
  const positions = new Float32Array(positionsCount * 3);

  const normalsCount = size;
  const normals = new Float32Array(normalsCount * 3);

  const uvsCount = size;
  const uvs = new Float32Array(uvsCount * 2);

  const temp = new Float32Array(3);
  let vertexIndex = 0;
  let curIndex = 0;

  for (let iy = 0; iy <= heightSegments; iy++) {
    let v = iy / heightSegments;
    let t = v * theta + thetaOffset;
    let cosTheta = Math.cos(t);
    let sinTheta = Math.sin(t);

    for (let ix = 0; ix <= widthSegments; ix++) {
      let u = ix / widthSegments;
      let p = u * phi + phiOffset;
      let cosPhi = Math.cos(p);
      let sinPhi = Math.sin(p);

      temp[0] = -cosPhi * sinTheta;
      temp[1] = -cosTheta;
      temp[2] = sinPhi * sinTheta;

      positions[vertexIndex * 3] = radius * temp[0];
      positions[vertexIndex * 3 + 1] = radius * temp[1];
      positions[vertexIndex * 3 + 2] = radius * temp[2];

      let length = Math.sqrt(temp[0] * temp[0] + temp[1] * temp[1] + temp[2] * temp[2]);
      let n = 1 / length;
      normals[vertexIndex * 3] = temp[0] * n;
      normals[vertexIndex * 3 + 1] = temp[1] * n;
      normals[vertexIndex * 3 + 2] = temp[2] * n;

      uvs[vertexIndex * 2] = u;
      uvs[vertexIndex * 2 + 1] = v;

      vertexIndex++;
    }

    if (iy > 0) {
      for (let i = vertexIndex - 2 * (widthSegments + 1); i + widthSegments + 2 < vertexIndex; i++) {
        const a = i;
        const b = i + 1;
        const c = i + widthSegments + 1;
        const d = i + widthSegments + 2;
        indices[curIndex] = a;
        indices[curIndex + 1] = b;
        indices[curIndex + 2] = c;

        indices[curIndex + 3] = c;
        indices[curIndex + 4] = b;
        indices[curIndex + 5] = d;

        curIndex += 6;
      }
    }
  }

  const mesh = world.createMesh({
    primitives: [
      {
        mode: WebSG.MeshPrimitiveMode.TRIANGLES,
        indices: world.createAccessorFrom(indices.buffer, {
          componentType: WebSG.AccessorComponentType.Uint16,
          count: indicesCount,
          type: WebSG.AccessorType.SCALAR,
        }),
        attributes: {
          POSITION: world.createAccessorFrom(positions.buffer, {
            componentType: WebSG.AccessorComponentType.Float32,
            count: positionsCount,
            type: WebSG.AccessorType.VEC3,
            dynamic: true,
          }),
          NORMAL: world.createAccessorFrom(normals.buffer, {
            componentType: WebSG.AccessorComponentType.Float32,
            count: normalsCount,
            type: WebSG.AccessorType.VEC3,
            normalized: true,
            dynamic: true,
          }),
          TEXCOORD_0: world.createAccessorFrom(uvs.buffer, {
            componentType: WebSG.AccessorComponentType.Float32,
            count: uvsCount,
            type: WebSG.AccessorType.VEC2,
          }),
        },
      },
    ],
  });

  return {
    vertexCount: vertexIndex,
    positions,
    mesh,
  };
}

let audioData;
let vertexCount;
let sphereMeshData;
let positionAccessor;
let sphereNode;
let beamMaterial;
let accNoise = 0;

world.onload = () => {
  const meshMaterial = world.createMaterial({
    baseColorFactor: [0, 0, 0, 1],
    metallicFactor: 0.5,
    roughnessFactor: 0.7,
  });

  sphereMeshData = generateSphereMesh(1, 32, 16, Math.PI, 0, Math.PI * 2, 0);

  positionAccessor = sphereMeshData.mesh.primitives[0].getAttribute("POSITION");

  vertexCount = sphereMeshData.vertexCount;

  sphereMeshData.mesh.primitives[0].material = meshMaterial;

  sphereNode = world.createNode({
    translation: [7, 8, -17],
    scale: [3, 3, 3],
    mesh: sphereMeshData.mesh,
  });

  world.environment.addNode(sphereNode);

  const beamMesh = world.findMeshByName("Tube_Light_01");
  beamMaterial = beamMesh.primitives[0].material;
  beamMaterial.emissiveFactor[0] = 1;
  beamMaterial.emissiveFactor[1] = 1;
  beamMaterial.emissiveFactor[2] = 1;

  const audioDataSize = thirdroom.getAudioDataSize();
  audioData = new Uint8Array(audioDataSize);
};

world.onupdate = (dt, elapsed) => {
  thirdroom.getAudioFrequencyData(audioData);

  let lowFreqAvg = 0;

  for (let i = 0; i < 128; i++) {
    lowFreqAvg += audioData[i];
  }

  lowFreqAvg = lowFreqAvg / 128 / 255;

  for (let i = 0; i < vertexCount; i++) {
    let x = sphereMeshData.positions[i * 3];
    let y = sphereMeshData.positions[i * 3 + 1];
    let z = sphereMeshData.positions[i * 3 + 2];

    let length = Math.sqrt(x * x + y * y + z * z);
    let n = 1 / length;
    x = x * n;
    y = y * n;
    z = z * n;

    const noise = 1 + 0.5 * improvedNoise(x * 5 + elapsed, y * 5 + elapsed, z * 5 + elapsed);

    sphereMeshData.positions[i * 3] = x * noise;
    sphereMeshData.positions[i * 3 + 1] = y * noise;
    sphereMeshData.positions[i * 3 + 2] = z * noise;

    accNoise += noise;
  }

  let scale = 3 * (1 + lowFreqAvg);
  sphereNode.scale[0] = scale;
  sphereNode.scale[1] = scale;
  sphereNode.scale[2] = scale;

  beamMaterial.emissiveFactor[0] = 1;
  beamMaterial.emissiveFactor[1] = 1 - lowFreqAvg;
  beamMaterial.emissiveFactor[2] = 1 - lowFreqAvg;

  positionAccessor.updateWith(sphereMeshData.positions.buffer);
};
